spring-core/src/main/java/org/springframework/core/io/buffer/NettyDataBuffer.java

```diff
@@ -229,6 +237,51 @@ public NettyDataBuffer write(ByteBuf... byteBufs) {
 		return this;
 	}
 
+	@Override
+	public DataBuffer write(CharSequence charSequence, Charset charset) {
+		Assert.notNull(charSequence, "'charSequence' must not be null");
+		Assert.notNull(charset, "'charset' must not be null");
+		if (StandardCharsets.UTF_8.equals(charset)) {
+			ByteBufUtil.writeUtf8(this.byteBuf, charSequence);
+		}
+		else if (StandardCharsets.US_ASCII.equals(charset)) {
+			ByteBufUtil.writeAscii(this.byteBuf, charSequence);
+		}
+		else {
+			CharsetEncoder charsetEncoder = charset.newEncoder()
+					.onMalformedInput(CodingErrorAction.REPLACE)
+					.onUnmappableCharacter(CodingErrorAction.REPLACE);
+			CharBuffer inBuffer = CharBuffer.wrap(charSequence);
+			int estimatedSize = (int) (inBuffer.remaining() * charsetEncoder.averageBytesPerChar());
+			this.byteBuf.ensureWritable(estimatedSize);
+			charsetEncoder.reset();
+			ByteBuffer outBuffer = asByteBuffer(this.byteBuf.writerIndex(), writableByteCount());
+			try {
+				for (; ; ) {
+					CoderResult cr = inBuffer.hasRemaining() ?
+							charsetEncoder.encode(inBuffer, outBuffer, true) : CoderResult.UNDERFLOW;
+					if (cr.isUnderflow()) {
+						cr = charsetEncoder.flush(outBuffer);
+					}
+					if (cr.isUnderflow()) {
+						break;
+					}
+					if (cr.isOverflow()) {
+						writePosition(outBuffer.position());
+						int maximumSize = (int) (inBuffer.remaining() * charsetEncoder.maxBytesPerChar());
+						this.byteBuf.ensureWritable(maximumSize);
+						outBuffer = asByteBuffer(this.byteBuf.writerIndex(), writableByteCount());
+					}
+				}
+			}
+			catch (CoderMalfunctionError ex) {
+				throw new IllegalStateException(ex);
+			}
+			writePosition(outBuffer.position());
+		}
+		return this;
```

Comment:
The non-UTF-8, non-ASCII code should really live in a base class, since it's shared between `DefaultDataBuffer` and `NettyDataBuffer`.  Or perhaps better yet: give the method a default implementation in `DataBuffer`.